[FEAT] Slight restructuring to allow for comments, better text handli…

dyst5422 opened the pull request:
…ng, and allowing more information to be passed through

Haven't gotten to the tests yet.

dyst5422 commented on src/lib.rs:
```diff
@@ -276,20 +297,39 @@ impl Element {
+                XMLNode::Element(elem) => Some(elem),
+                _ => None,
+            })
+            .find(|e| k.match_element(e))
```
should probably provide a way to get at text, commen t, etc nodes

eminence replied:
Agreed 👍

eminence commented on src/lib.rs:
```diff
@@ -161,21 +169,21 @@ impl Element {
 
     /// Parses some data into an Element
-    pub fn parse<R: Read>(r: R) -> Result<Element, ParseError> {
+    pub fn parse<R: Read>(r: R) -> Result<Vec<XMLNode>, ParseError> {
```
If at all possible, I'd like to keep the original signature and return only a single `Element`, and not a `Vec<XMLNode>`.

eminence commented on src/lib.rs:
```diff
@@ -161,21 +169,21 @@ impl Element {
     /// Parses some data into an Element
-    pub fn parse<R: Read>(r: R) -> Result<Element, ParseError> {
+    pub fn parse<R: Read>(r: R) -> Result<Vec<XMLNode>, ParseError> {
         let mut reader = EventReader::new(r);
```
Instead of `EventReader::new`, we'll need to use `EventReader::new_with_config`, since the default configuration will ignore comments)

dyst5422 commented:
I like that solution. Yeah, I wasn’t so hot on the different parse signature myself, but wanted to ensure completeness in what is returned. I think `parse_all` satisfies both.

dyst5422 commented:
Just added `parse_all` and reverted the signature for `parse`. Additionally, all tests pass and a test was added checking that `parse_all` returns the proper number of nodes.

eminence commented on src/lib.rs:
```diff
@@ -191,25 +198,41 @@ impl Element {
-                | Ok(XmlEvent::ProcessingInstruction { .. }) => {
-                    return Err(ParseError::CannotParse)
+                Ok(XmlEvent::Comment(comment_string)) => {
+                    println!("comment");
```
Debugging code here to be removed

dyst5422 replied:
Jumped on it right as I found that error!

eminence commented:
Thanks!  I've merged this, plus added a few new features.  This has been published as `v0.10.0`
